File: perceptron.cc
#include "perceptron.h"
#include <vector>
using std::vector;

Perceptron::Perceptron(int n_features)
      : learned_weights_(n_features),
        epoch_(0),
        learn_rate_(1) {
}

Perceptron::Perceptron(int n_features, int learn_rate)
      : learned_weights_(n_features),
        epoch_(0),
        learn_rate_(learn_rate) {
}

Perceptron::Perceptron(const Perceptron& other)
      : learned_weights_(other.learned_weights_),
        epoch_(other.epoch_),
        learn_rate_(other.learn_rate_) {
}

Perceptron::~Perceptron() {
}

void Perceptron::EpochTick() {
   ++epoch_;
}

int Perceptron::GetEpoch() const {
   return epoch_;
}

void Perceptron::ResetEpoch() {
   epoch_ = 0;
}

double Perceptron::GetActivation(const vector<double>& features) const {
   double activation = 0;
   for (int i = 0; i < features.size() && i < learned_weights_.size(); ++i) {
      activation += features[i] * learned_weights_[i];
   }
   return activation;
}

bool Perceptron::GiveTrainingData(const vector<double>& features, bool label) {
   bool sgn = GetActivation(features) > 0;
   double step = Alpha() * ((label ? 1 : -1) - (sgn ? 1 : -1));
   for (int i = 0; i < learned_weights_.size() && i < features.size(); ++i) {
      learned_weights_[i] += step * features[i];
   }
   return label == sgn;
}

double Perceptron::Alpha() const {
   return learn_rate_ / (learn_rate_ + epoch_);
}

File: perceptron.py
from __future__ import division
import numpy as np

class Perceptron:
   def __init__(self, n_features, learn_rate=1):
      self.learned_weights = np.random.randn(n_features)
      self.epoch = 0
      self.learn_rate = learn_rate

   def epoch_tick(self):
      self.epoch += 1

   def get_epoch(self):
      return self.epoch

   def reset_epoch(self):
      self.epoch = 0

   def get_activation(self, features):
      return np.dot(self.learned_weights, features)

   def classify(self, features):
      return self.get_activation(features) > 0

   def alpha(self):
      return self.learn_rate / (self.learn_rate + self.epoch)

   def give_training_data(self, features, label):
      act = self.get_activation(features)
      sgn = 1 if act > 0 else -1
      step = self.alpha() * ((1 if bool(label) else -1) - sgn)
      self.learned_weights += step * features
      return bool(act > 0) == bool(label)

